toolkit/tools/pkg/imagecustomizerlib/cosicommon.go

```diff
@@ -188,5 +217,232 @@ func populateMetadata(data ImageBuildData) error {
+	}
+
+	verityFile.Sha384 = veritySha384
+
+	verityFile.UncompressedSize = uint64(verityStat.Size())
+
+	return nil
+}
+
+func extractPartitionsForCosi(diskPartitions []diskutils.PartitionInfo, outDir string, basename string, partitionFormat string, imageUuid [UuidSize]byte, roothash string, rootPartUUID string, roothashFilepath string) ([]outputPartitionMetadata, error) {
+	// Stores the output partition metadata that will be written to JSON file
+	var partitionMetadataOutput []outputPartitionMetadata
+
+	// Extract partitions to files
+	for _, partition := range diskPartitions {
+		if partition.Type != "part" {
+			continue
+		}
+
+		partitionNum, err := getPartitionNum(partition.Path)
+		if err != nil {
+			return nil, err
+		}
+
+		partitionFilename := basename + "_" + strconv.Itoa(partitionNum)
+		rawFilename := partitionFilename + ".raw"
+
+		partitionFilepath, err := copyBlockDeviceToFile(outDir, partition.Path, rawFilename)
+		if err != nil {
+			return nil, err
+		}
+
+		partitionFullFilePath, err := filepath.Abs(partitionFilepath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get absolute path (%s):\n%w", partitionFilepath, err)
+		}
+
+		// Sanity check the partition file
+		err = checkFileSystemFile(partition.FileSystemType, partitionFullFilePath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to check file system integrity (%s):\n%w", partitionFilepath, err)
+		}
+
+		// Get uncompressed size for raw files
+		var uncompressedPartitionFileSize uint64
+		stat, err := os.Stat(partitionFullFilePath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to stat raw file %s: %w", partitionFilepath, err)
+		}
+		uncompressedPartitionFileSize = uint64(stat.Size())
+
+		// Handle partition formats
+		switch partitionFormat {
+		case "raw":
+			// Do nothing for "raw" case
+		case "raw-zst":
+			partitionFilepath, err = extractRawZstPartition(partitionFilepath, imageUuid, partitionFilename, outDir)
+			if err != nil {
+				return nil, err
+			}
+		default:
+			return nil, fmt.Errorf("unsupported partition format (supported: raw, raw-zst): %s", partitionFormat)
+		}
+
+		// If this is the roothash partition, skip further processing as we will assign its information to root partition
+		if partition.FileSystemType == "DM_verity_hash" {
+			logger.Log.Infof("Skipping metadata construction for roothash partition: %s", partition.Path)
+			continue
+		}
+
+		// Construct metadata
+		partitionMetadata, err := constructOutputPartitionMetadata(partition, partitionNum, partitionFilepath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to construct partition metadata:\n%w", err)
+		}
+		partitionMetadata.UncompressedSize = uncompressedPartitionFileSize
+
+		// Assign verity information to the root partition
+		if roothash != "" && rootPartUUID != "" && roothashFilepath != "" && partition.PartUuid == rootPartUUID {
+			roothashFilepath, err = extractRawZstPartition(roothashFilepath, imageUuid, partitionFilename, outDir)
```

Comment:
This code seems odd. Why are you extracting the verity hash partition here? Why not do this in 2 phases: 1. Extract all the partitions and their metadata, 2. Convert the flat list of partition metadata into the tree structure wanted by COSI?